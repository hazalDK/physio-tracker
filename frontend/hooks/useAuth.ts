import * as SecureStore from "expo-secure-store";
import axios from "axios";
import { useNavigation } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import { RootStackParamsList } from "@/types/navigation";
import { Alert } from "react-native";

export function useAuth() {
  const navigation =
    useNavigation<StackNavigationProp<RootStackParamsList, "login">>();

  const getToken = async () => {
    return await SecureStore.getItemAsync("access_token");
  };

  const refreshToken = async () => {
    try {
      const refreshToken = await SecureStore.getItemAsync("refresh_token");
      if (!refreshToken) {
        throw new Error("No refresh token available");
      }

      const refreshUrl = process.env.API_URL || "http://192.168.68.111:8000";
      const refreshResponse = await axios.post(
        `${refreshUrl}/api/token/refresh/`,
        { refresh: refreshToken }
      );

      const newToken = refreshResponse.data.access;
      const newRefreshToken = refreshResponse.data.refresh;

      await Promise.all([
        SecureStore.setItemAsync("access_token", newToken),
        SecureStore.setItemAsync("refresh_token", newRefreshToken),
      ]);

      return newToken;
    } catch (error) {
      await Promise.all([
        SecureStore.deleteItemAsync("access_token"),
        SecureStore.deleteItemAsync("refresh_token"),
      ]);

      Alert.alert("Session Expired", "Please login again", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);

      return null;
    }
  };

  const createApiInstance = async () => {
    let token = await getToken();

    if (!token) {
      Alert.alert("Login Required", "Please sign in to continue", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);
      return null;
    }

    return axios.create({
      baseURL: process.env.API_URL || "http://192.168.68.111:8000",
      timeout: 10000,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });
  };

  return { getToken, refreshToken, createApiInstance };
}
