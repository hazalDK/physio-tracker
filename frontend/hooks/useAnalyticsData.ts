import { useState, useEffect } from "react";
import axios from "axios";
import { useAuth } from "./useAuth";

// Helper function to format date
const formatDate = (date: Date) => {
  return date.toISOString().split("T")[0];
};

// Type for analytics data (can be either adherence or pain data)
interface ChartData {
  labels: string[];
  datasets: {
    data: number[];
  }[];
}

/**
 * Custom hook to fetch and manage analytics data (adherence or pain levels).
 * @param {string} endpoint - The API endpoint to fetch data from.
 * @param {"adherence" | "pain"} dataType - The type of data to fetch (either adherence or pain).
 * @returns {Object} - An object containing loading state, chart data, average, history, error, and functions to navigate weeks.
 * @throws {Error} - Throws an error if the API call fails.
 * */
export function useAnalyticsData<T>(
  endpoint: string,
  dataType: "adherence" | "pain"
): object {
  const [loading, setLoading] = useState(true);
  const [chartData, setChartData] = useState<ChartData>({
    labels: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
    datasets: [{ data: [0, 0, 0, 0, 0, 0, 0] }],
  });
  const [average, setAverage] = useState(0);
  const [history, setHistory] = useState<T[]>([]);
  const [error, setError] = useState<string | null>(null);

  // Add week navigation state
  const [currentEndDate, setCurrentEndDate] = useState(new Date());
  const [weekTitle, setWeekTitle] = useState("Current Week");
  const { createApiInstance, refreshToken } = useAuth();

  // Function to go to previous week
  const goToPreviousWeek = () => {
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() - 7);
    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Function to go to next week (but not beyond current date)
  const goToNextWeek = () => {
    const today = new Date();
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() + 7);

    // Don't go beyond current date
    if (newEndDate > today) {
      newEndDate.setTime(today.getTime());
    }

    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Update week title based on date range
  const updateWeekTitle = (endDate: Date) => {
    const startDate = new Date(endDate);
    startDate.setDate(startDate.getDate() - 6);

    const today = new Date();
    const isCurrentWeek =
      today.getDate() === endDate.getDate() &&
      today.getMonth() === endDate.getMonth() &&
      today.getFullYear() === endDate.getFullYear();

    if (isCurrentWeek) {
      setWeekTitle("Current Week");
    } else {
      const startMonth = startDate.toLocaleString("default", {
        month: "short",
      });
      const endMonth = endDate.toLocaleString("default", { month: "short" });

      if (startMonth === endMonth) {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endDate.getDate()}`
        );
      } else {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endMonth} ${endDate.getDate()}`
        );
      }
    }
  };

  // Function to fetch data for a specific end date
  const fetchData = async (endDate: Date) => {
    setLoading(true);
    try {
      const api = await createApiInstance();
      if (!api) return;

      try {
        // Fetch stats from the API with date parameter
        const response = await api.get(
          `${endpoint}?end_date=${formatDate(endDate)}`
        );
        if (response.data) {
          setChartData(response.data.chart_data);

          if (dataType === "adherence") {
            setAverage(response.data.average_adherence);
          } else {
            setAverage(response.data.average_pain);
          }

          setHistory(response.data.history);
        }
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();
          if (!newToken) return;

          const api = await createApiInstance();
          if (!api) return;

          // Retry with new token
          const response = await api.get(
            `${endpoint}?end_date=${formatDate(endDate)}`
          );

          if (response.data) {
            setChartData(response.data.chart_data);

            if (dataType === "adherence") {
              setAverage(response.data.average_adherence);
            } else {
              setAverage(response.data.average_pain);
            }

            setHistory(response.data.history);
          }
        } else {
          // Handle other errors
          console.error("API request failed:", error);
          setError(
            `Failed to load your ${
              dataType === "adherence" ? "exercises" : "pain data"
            }. Please check your connection and try again.`
          );
        }
      }
    } catch (error) {
      console.error(
        `Failed to load your ${
          dataType === "adherence" ? "exercises" : "pain data"
        }.`
      );
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    // Initial data fetch for current week
    fetchData(currentEndDate);
    updateWeekTitle(currentEndDate);
  }, []);

  return {
    loading,
    chartData,
    average,
    history,
    error,
    weekTitle,
    goToPreviousWeek,
    goToNextWeek,
  };
}
