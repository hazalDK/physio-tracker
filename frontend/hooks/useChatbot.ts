import { useState, useEffect, useCallback } from "react";
import { Alert } from "react-native";
import Toast from "react-native-root-toast";
import { messageType } from "@/types/chatbot";
import { createChatApi, apiRequestWithRefresh } from "../api";
import { useAuth } from "./useAuth";

export default function useChatbot() {
  const auth = useAuth();
  const [messages, setMessages] = useState<messageType[]>([]);
  const [userExercises, setUserExercises] = useState([]);
  const [recentPain, setRecentPain] = useState(0);
  const [isLoading, setIsLoading] = useState(false);
  const [isInitialLoading, setIsInitialLoading] = useState(true);
  const [api, setApi] = useState<any>(null);

  // Initialize API and chat
  useEffect(() => {
    const initializeApi = async () => {
      const apiInstance = await auth.createApiInstance();
      setApi(apiInstance);
    };

    initializeApi();

    // Set initial messages
    setMessages([
      {
        id: "disclaimer",
        text: "DISCLAIMER",
        sender: "disclaimer",
      },
      {
        id: "welcome",
        text: "Welcome to your physiotherapy assistant! How can I help you today?",
        sender: "bot",
      },
    ]);
  }, []);

  // Fetch user data when API is available
  useEffect(() => {
    if (api) {
      fetchUserData();
    }
  }, [api]);

  const fetchUserData = async () => {
    if (!api) return;

    setIsInitialLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const fetchData = async () => {
        const [reportsData, exercisesData] = await Promise.all([
          chatApi.fetchReports(),
          chatApi.fetchUserExercises(),
        ]);

        return { reportsData, exercisesData };
      };

      const { reportsData, exercisesData } = await apiRequestWithRefresh(
        fetchData,
        auth
      );

      setUserExercises(exercisesData);

      if (reportsData.length > 0) {
        setRecentPain(reportsData[0].pain_level);
      }
    } catch (error) {
      console.error("Failed to fetch user data:", error);

      Alert.alert(
        "Error",
        "Failed to load your exercises. Please check your connection and try again."
      );
    } finally {
      setIsInitialLoading(false);
    }
  };

  const sendMessage = async (inputText: string) => {
    if (!inputText.trim() || !api) return;

    // Add user message to chat
    const userMessageId = Date.now().toString();
    setMessages((prev) => [
      ...prev,
      { id: userMessageId, text: inputText, sender: "user" },
    ]);

    setIsLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Create context object with user's exercise data
      const exerciseContext = {
        exercises: userExercises,
        recentPain: recentPain,
      };

      // Use the wrapper function for API calls with refresh
      const sendMessageCall = async () => {
        return await chatApi.sendChatMessage(inputText, exerciseContext);
      };

      // Send message to API with refresh handling
      const response = await apiRequestWithRefresh(sendMessageCall, auth);

      // Add bot response to chat
      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: response.message,
          sender: "bot",
        },
      ]);
    } catch (error) {
      console.error("Error sending message:", error);

      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: "Sorry, I encountered an error. Please try again.",
          sender: "bot",
        },
      ]);
    } finally {
      setIsLoading(false);
    }
  };

  const resetChat = async () => {
    if (!api) return;

    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const resetCall = async () => {
        await chatApi.resetChat();
      };

      await apiRequestWithRefresh(resetCall, auth);

      // Reset messages
      setMessages([
        {
          id: "disclaimer",
          text: "DISCLAIMER",
          sender: "disclaimer",
        },
        {
          id: "welcome",
          text: "Welcome to your physiotherapy assistant! How can I help you today?",
          sender: "bot",
        },
      ]);

      // Refresh user data
      fetchUserData();

      Toast.show("Chat reset!", {
        duration: Toast.durations.SHORT,
        position: Toast.positions.BOTTOM,
        backgroundColor: "#14b8a6",
        textColor: "white",
        shadow: true,
      });
    } catch (error) {
      console.error("Failed to reset chat:", error);
      Alert.alert("Error", "Could not reset chat.");
    }
  };

  return {
    messages,
    isLoading,
    isInitialLoading,
    sendMessage,
    resetChat,
  };
}
