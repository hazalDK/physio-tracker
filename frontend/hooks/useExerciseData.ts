import { useState, useEffect } from "react";
import { Alert } from "react-native";
import axios from "axios";
import { ExerciseItem, UserExerciseItem } from "@/types/exercise";
import { useAuth } from "./useAuth";

export function useExerciseData(
  exerciseId: number,
  userExerciseId: number | null
) {
  const [loading, setLoading] = useState(true);
  const [exercise, setExercise] = useState<ExerciseItem>();
  const [userExercise, setUserExercise] = useState<UserExerciseItem>();
  const { createApiInstance, refreshToken } = useAuth();

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      try {
        const api = await createApiInstance();
        if (!api) return;

        try {
          const [userExercisesResponse, exercisesResponse] = await Promise.all([
            api.get(`/user-exercises/${userExerciseId}/`),
            api.get(`/exercises/${exerciseId}/`),
          ]);

          setUserExercise(
            !Array.isArray(userExercisesResponse.data)
              ? userExercisesResponse.data
              : ({} as UserExerciseItem)
          );

          setExercise(
            !Array.isArray(exercisesResponse.data)
              ? exercisesResponse.data
              : ({} as ExerciseItem)
          );
        } catch (error) {
          if (axios.isAxiosError(error) && error.response?.status === 401) {
            const newToken = await refreshToken();
            if (!newToken) return;

            const api = await createApiInstance();
            if (!api) return;

            const [retryUserExercises, retryExercises] = await Promise.all([
              api.get(`/user-exercises/${userExerciseId}/`),
              api.get(`/exercises/${exerciseId}/`),
            ]);

            setUserExercise(
              !Array.isArray(retryUserExercises.data)
                ? retryUserExercises.data
                : ({} as UserExerciseItem)
            );

            setExercise(
              !Array.isArray(retryExercises.data)
                ? retryExercises.data
                : ({} as ExerciseItem)
            );
          } else {
            console.error("API request failed:", error);
            Alert.alert(
              "Error",
              "Failed to load your exercises. Please check your connection and try again."
            );
            setExercise({} as ExerciseItem);
            setUserExercise({} as UserExerciseItem);
          }
        }
      } catch (error) {
        console.error("Failed to load exercise data:", error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [exerciseId, userExerciseId]);

  return { loading, exercise, userExercise };
}
