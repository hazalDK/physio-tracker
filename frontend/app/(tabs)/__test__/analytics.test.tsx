import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";
import Analytics from "../analytics";
import { useAnalyticsData } from "@/hooks/useAnalyticsData";

// Mock the dependencies
jest.mock("@/hooks/useAnalyticsData");

// Mock useFocusEffect to prevent infinite loops
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useFocusEffect: jest.fn((callback) => callback()),
  };
});

// Mock expo-font
jest.mock("expo-font", () => ({
  isLoaded: jest.fn(() => true),
  loadAsync: jest.fn(() => Promise.resolve()),
  __internal__: {
    getAssetForModule: jest.fn(() => 1),
  },
  Font: {
    isLoaded: jest.fn(() => true),
    loadAsync: jest.fn(() => Promise.resolve()),
  },
}));

jest.mock("@expo/vector-icons", () => {
  const React = require("react");
  const mockIcon = (props) => React.createElement("Icon", props);

  // Return an object with all common icon sets
  return {
    Entypo: mockIcon,
    Ionicons: mockIcon,
    FontAwesome: mockIcon,
    MaterialIcons: mockIcon,
    MaterialCommunityIcons: mockIcon,
    // Add other icon sets you're using
    createIconSet: jest.fn(() => mockIcon),
  };
});

jest.mock("@react-navigation/material-top-tabs", () => {
  const { View } = require("react-native");
  return {
    createMaterialTopTabNavigator: () => ({
      Navigator: ({ children }) => <View>{children}</View>,
      Screen: ({ component: Component }) => <Component />,
    }),
  };
});

jest.mock("react-native-chart-kit", () => ({
  LineChart: () => "LineChart",
  BarChart: () => "BarChart",
}));

jest.mock("@/components/ExerciseHistoryDetail", () => "ExerciseHistoryDetail");

describe("Analytics Component", () => {
  // Mock data for adherence test
  const mockAdherenceData = {
    loading: false,
    chartData: {
      labels: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
      datasets: [{ data: [75, 80, 100, 85, 90, 70, 95] }],
    },
    average: 85,
    history: [
      { date: "2025-04-23", completed: 3, adherence: 75 },
      { date: "2025-04-24", completed: 4, adherence: 100 },
    ],
    error: null,
    weekTitle: "April 20 - April 26, 2025",
    goToPreviousWeek: jest.fn(),
    goToNextWeek: jest.fn(),
    refreshData: jest.fn(),
  };

  // Mock data for pain test
  const mockPainData = {
    loading: false,
    chartData: {
      labels: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
      datasets: [{ data: [5, 4, 3, 4, 2, 3, 2] }],
    },
    average: 3.3,
    history: [
      { date: "2025-04-23", exercises: 3, pain_level: 4 },
      { date: "2025-04-24", exercises: 4, pain_level: 2 },
    ],
    error: null,
    weekTitle: "April 20 - April 26, 2025",
    goToPreviousWeek: jest.fn(),
    goToNextWeek: jest.fn(),
    refreshData: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders AdherenceGraph correctly with data", () => {
    (useAnalyticsData as jest.Mock).mockReturnValue(mockAdherenceData);

    const { getByText, getByTestId } = render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    expect(getByText("Average Adherence: 85%")).toBeTruthy();
    expect(getByText("Weekly Adherence")).toBeTruthy();
    expect(getByText("Exercise History")).toBeTruthy();
    expect(getByTestId("week-title")).toHaveTextContent(
      "April 20 - April 26, 2025"
    );
  });

  it("renders PainLevelGraph correctly with data", () => {
    (useAnalyticsData as jest.Mock).mockReturnValue(mockPainData);

    const { getByText, getByTestId } = render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    expect(getByText("Average Pain Level: 3.3/10")).toBeTruthy();
    expect(getByText("Weekly Pain Levels")).toBeTruthy();
    expect(getByText("Pain History")).toBeTruthy();
    expect(getByTestId("pain-week-title")).toHaveTextContent(
      "April 20 - April 26, 2025"
    );
  });

  it("shows loading indicator when data is loading for adherence", () => {
    console.log("Mocking loading state for adherence data");
    (useAnalyticsData as jest.Mock).mockReturnValue({
      ...mockAdherenceData,
      loading: true,
    });

    const { getByTestId } = render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    // Note: You need to add testID="loading-indicator" to the ActivityIndicator
    expect(getByTestId("loading-indicator")).toBeTruthy();
  });

  it("shows error message when there is an error for adherence", () => {
    const errorMessage = "Failed to load data";
    (useAnalyticsData as jest.Mock).mockReturnValue({
      ...mockAdherenceData,
      error: errorMessage,
    });

    const { getByTestId } = render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    expect(getByTestId("adherence-error-message")).toBeTruthy();
  });

  it("shows error message when there is an error for pain level", () => {
    const errorMessage = "Failed to load data";
    (useAnalyticsData as jest.Mock).mockReturnValue({
      ...mockPainData,
      error: errorMessage,
    });

    const { getByTestId } = render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    expect(getByTestId("pain-error-message")).toBeTruthy();
  });

  it("navigation controls work correctly for adherence", () => {
    (useAnalyticsData as jest.Mock).mockReturnValue(mockAdherenceData);

    const { getByTestId } = render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    // Note: You need to add testID attributes to these buttons
    fireEvent.press(getByTestId("prev-week-button"));
    expect(mockAdherenceData.goToPreviousWeek).toHaveBeenCalledTimes(1);

    fireEvent.press(getByTestId("next-week-button"));
    expect(mockAdherenceData.goToNextWeek).toHaveBeenCalledTimes(1);
  });

  it("refreshes data on focus for pain level", () => {
    (useAnalyticsData as jest.Mock).mockReturnValue(mockPainData);

    render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    expect(mockPainData.refreshData).toHaveBeenCalledTimes(2);
  });

  it("refreshes data on focus for adherence", () => {
    (useAnalyticsData as jest.Mock).mockReturnValue(mockAdherenceData);

    render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    expect(mockAdherenceData.refreshData).toHaveBeenCalledTimes(2);
  });

  it("shows loading indicator when pain level data is loading", () => {
    console.log("Mocking loading state for pain level data");
    (useAnalyticsData as jest.Mock).mockReturnValue({
      ...mockPainData,
      loading: true,
    });

    const { getByTestId } = render(
      <NavigationContainer>
        <Analytics />
      </NavigationContainer>
    );

    expect(getByTestId("pain_level_loading-indicator")).toBeTruthy();
  });
});
