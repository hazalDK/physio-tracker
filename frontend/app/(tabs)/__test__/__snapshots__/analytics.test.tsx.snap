// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics Component renders correctly with data 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(255, 255, 255, 1)",
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <View>
    <RCTScrollView
      style={
        {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingLeft": 16,
          "paddingRight": 16,
        }
      }
    >
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 16,
              "marginTop": 128,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="prev-week-button"
          >
            <Icon
              color="#14b8a6"
              name="chevron-back"
              size={24}
            />
          </View>
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "700",
                "lineHeight": 28,
              }
            }
            testID="week-title"
          >
            April 20 - April 26, 2025
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="next-week-button"
          >
            <Icon
              color="#14b8a6"
              name="chevron-forward"
              size={24}
            />
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "marginTop": 16,
            }
          }
        >
          <View
            style={
              {
                "borderBottomColor": "rgba(209, 213, 219, 1)",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderBottomWidth": 2,
                "borderLeftColor": "rgba(209, 213, 219, 1)",
                "borderLeftWidth": 2,
                "borderRightColor": "rgba(209, 213, 219, 1)",
                "borderRightWidth": 2,
                "borderTopColor": "rgba(209, 213, 219, 1)",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderTopWidth": 2,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 18,
                  "fontWeight": "700",
                  "lineHeight": 28,
                  "textAlign": "center",
                }
              }
            >
              Average Adherence: 
              85
              %
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "marginTop": 24,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "700",
                "lineHeight": 28,
                "marginBottom": 8,
              }
            }
          >
            Weekly Adherence
          </Text>
          LineChart
        </View>
        <View
          style={
            {
              "marginBottom": 32,
              "marginTop": 24,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "700",
                "lineHeight": 28,
                "marginBottom": 8,
              }
            }
          >
            Exercise History
          </Text>
          <View
            style={
              {
                "borderBottomColor": "rgba(209, 213, 219, 1)",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderBottomWidth": 2,
                "borderLeftColor": "rgba(209, 213, 219, 1)",
                "borderLeftWidth": 2,
                "borderRightColor": "rgba(209, 213, 219, 1)",
                "borderRightWidth": 2,
                "borderTopColor": "rgba(209, 213, 219, 1)",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderTopWidth": 2,
                "marginBottom": 12,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "700",
                }
              }
              testID="adherence-history-date-0"
            >
              Sunday, April 27
            </Text>
            <Text>
              3/4
               active exercises completed
            </Text>
            <Text
              style={
                {
                  "color": "#14b8a6",
                }
              }
            >
              Adherence: 
              75
              %
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "rgba(209, 213, 219, 1)",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderBottomWidth": 2,
                "borderLeftColor": "rgba(209, 213, 219, 1)",
                "borderLeftWidth": 2,
                "borderRightColor": "rgba(209, 213, 219, 1)",
                "borderRightWidth": 2,
                "borderTopColor": "rgba(209, 213, 219, 1)",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderTopWidth": 2,
                "marginBottom": 12,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "700",
                }
              }
              testID="adherence-history-date-1"
            >
              Saturday, April 26
            </Text>
            <Text>
              4/4
               active exercises completed
            </Text>
            <Text
              style={
                {
                  "color": "#14b8a6",
                }
              }
            >
              Adherence: 
              100
              %
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <RCTScrollView
      style={
        {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingLeft": 16,
          "paddingRight": 16,
        }
      }
    >
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 16,
              "marginTop": 128,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="pain_level_prev_week_button"
          >
            <Icon
              color="#14b8a6"
              name="chevron-back"
              size={24}
            />
          </View>
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "700",
                "lineHeight": 28,
              }
            }
            testID="pain-week-title"
          >
            April 20 - April 26, 2025
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="pain_level_next_week_button"
          >
            <Icon
              color="#14b8a6"
              name="chevron-forward"
              size={24}
            />
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "marginTop": 16,
            }
          }
        >
          <View
            style={
              {
                "borderBottomColor": "rgba(209, 213, 219, 1)",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderBottomWidth": 2,
                "borderLeftColor": "rgba(209, 213, 219, 1)",
                "borderLeftWidth": 2,
                "borderRightColor": "rgba(209, 213, 219, 1)",
                "borderRightWidth": 2,
                "borderTopColor": "rgba(209, 213, 219, 1)",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderTopWidth": 2,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 18,
                  "fontWeight": "700",
                  "lineHeight": 28,
                  "textAlign": "center",
                }
              }
            >
              Average Pain Level: 
              85
              /10
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "marginTop": 24,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "700",
                "lineHeight": 28,
                "marginBottom": 8,
                "textAlign": "center",
              }
            }
          >
            Weekly Pain Levels
          </Text>
          BarChart
        </View>
        <View
          style={
            {
              "marginBottom": 32,
              "marginTop": 24,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "700",
                "lineHeight": 28,
                "marginBottom": 8,
              }
            }
          >
            Pain History
          </Text>
          <View
            style={
              {
                "borderBottomColor": "rgba(209, 213, 219, 1)",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderBottomWidth": 2,
                "borderLeftColor": "rgba(209, 213, 219, 1)",
                "borderLeftWidth": 2,
                "borderRightColor": "rgba(209, 213, 219, 1)",
                "borderRightWidth": 2,
                "borderTopColor": "rgba(209, 213, 219, 1)",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderTopWidth": 2,
                "marginBottom": 12,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "700",
                }
              }
              testID="pain-history-date-0"
            >
              Sunday, April 27
            </Text>
            <Text>
               exercises completed
            </Text>
            <Text
              style={
                {
                  "color": "rgba(239, 68, 68, 1)",
                }
              }
            >
              Pain Level: 
              /10
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "rgba(209, 213, 219, 1)",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderBottomWidth": 2,
                "borderLeftColor": "rgba(209, 213, 219, 1)",
                "borderLeftWidth": 2,
                "borderRightColor": "rgba(209, 213, 219, 1)",
                "borderRightWidth": 2,
                "borderTopColor": "rgba(209, 213, 219, 1)",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderTopWidth": 2,
                "marginBottom": 12,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "700",
                }
              }
              testID="pain-history-date-1"
            >
              Saturday, April 26
            </Text>
            <Text>
               exercises completed
            </Text>
            <Text
              style={
                {
                  "color": "rgba(239, 68, 68, 1)",
                }
              }
            >
              Pain Level: 
              /10
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <ExerciseHistoryDetail />
  </View>
</View>
`;
